### ISSUE TITLE
Implement multi-threading

### ISSUE THREAD

**Issue by vig42** [reposense/RepoSense#1452]
**What feature(s) would you like to see in RepoSense?**

Currently, RepoSense performs cloning and analysis in parallel; i.e. while Repo 1 is being analyzed, Repo 2 is being cloned. When we are running RepoSense on a large number of repos, we could achieve better performance if we clone and analyse more repos simultaneously using multi-threading.

**If possible, describe the solution**

We could implement multi-threading using Java’s ExecutorService and thread pooling. Each thread could do 3 things:

1.	Clone a single repo
2.	Perform analysis on that repo and generate the outputs 
3.	Delete all the data associated with that repo

Each repo has its own distinct folder to store the cloned repo data and a distinct folder to store the generated outputs (the commits and authorship JSON files). Since each thread handles 1 repo, there should be no conflicts with the file system between the threads. Furthermore, as there is no shared data modified by the threads, there should be no issue of race conditions.

**Performance improvement**

I made a rough implementation of this in #1451 as a proof of concept to test out the performance.

I tested it with 2 of the existing demos on the RepoSense websites: the CS2113 AY2019/20 Sem 2 individual project (72 repos) and team project (17 repos). With 4 threads, the time taken is cut down significantly.

For the 17 group project repos, the time taken to generate the report on my local machine went down from around 2 min 40 sec to 1 min 20 sec. For the 72 individual project repos, the time taken to generate the report went down from around 5 min 45 sec to 2 min 30 sec. This means the time taken is reduced by about 50-55% in each case.

Prof Damith had previously mentioned in [this comment](https://github.com/reposense/RepoSense/issues/540#issuecomment-460142643) on issue #540 that threading should not be considered unless there is significant time saving of 20%+, due to issues with debugging. However I feel that the potential savings of around 50% may make this viable.

**Comment by damithc** [comment]
> Prof Damith had previously mentioned in [this comment](https://github.com/reposense/RepoSense/issues/540#issuecomment-460142643) on issue #540 that threading should not be considered unless there is significant time saving of 20%+, due to issues with debugging. However I feel that the potential savings of around 50% may make this viable.

Yes, a 50% reduction is certainly not trivial. I wonder if we'll see a similar improvement when the report is generated in a CI environment, which is the more common case.

**Comment by gerhean** [comment]
Have you tested how much time is spent cloning and downloading the repos vs analyzing the code? Since they have different bottlenecks (cloning is io heavy and analyzing is cpu heavy) and it might be better to parallelize them differently.

**Comment by vig42** [comment]
> 
> 
> Have you tested how much time is spent cloning and downloading the repos vs analyzing the code? Since they have different bottlenecks (cloning is io heavy and analyzing is cpu heavy) and it might be better to parallelize them differently or separately.

Agreed, this definitely makes more sense, since the bottleneck resource is different for each of the 2 steps.

I did not do this initially since I was unsure about how to "schedule" the threads to ensure that the analysis for each repo would be done after it is cloned. Since then I have found a way to implement this using the CompletableFuture class. I also implemented [@dcshzj's suggestion](https://github.com/reposense/RepoSense/pull/1451#discussion_r575692572); I kept 4 threads available for cloning and set the number of threads for analysis to equal the number of CPU threads.

With this implementation, the time taken to produce the report for the CS2113 AY2019/20 Sem 2 individual project demo (72 repos) became even faster. It took only about 1 min 40 sec on my local machine, relative to around 5 min with the master branch.
